### PR TITLE
Added Dictionary Support

### PR DESCRIPTION
**Pull request by roee88**

### CODE REVIEW

**Review by cdoron — COMMENTED** [review]
Not specific to this PR, but the ffi README is still missing the following line:
`javah -cp ../../src/main/java org.apache.arrow.ffi.jni.JniWrapper`
between the `cmake` and the `make`

**Comment by roee88** [comment]
> Not specific to this PR, but the ffi README is still missing the following line:
> `javah -cp ../../src/main/java org.apache.arrow.ffi.jni.JniWrapper`
> between the `cmake` and the `make`

I didn't need to run this on my machine to get it working but maybe I changed things since the first compile and broke it. I will check.

File: java/ffi/src/main/java/org/apache/arrow/ffi/ArrayImporter.java [hunk 1]
```diff
@@ -29,23 +30,31 @@
...
-  ReferenceManager referenceManager;
+  private final BufferAllocator allocator;
+  private final FieldVector vector;
+  private final DictionaryProvider dictionaryProvider;
+
+  private ReferenceManager referenceManager;
   private int recursionLevel;
 
-  ArrayImporter(FieldVector vector) {
+  ArrayImporter(BufferAllocator allocator, FieldVector vector, DictionaryProvider dictionaryProvider) {
```
[the remark this answers is not in the document]
Comment (roee88) [reply java/ffi/src/main/java/org/apache/arrow/ffi/ArrayImporter.java:51]: Feel free to commit directly with your suggested changes

File: java/ffi/src/main/java/org/apache/arrow/ffi/SchemaExporter.java [hunk 2]
```diff
@@ -61,17 +65,12 @@ public void close() {
     }
   }
 
-  private final ArrowSchema schema;
-
-  SchemaExporter(ArrowSchema dst) {
```
[the remark this answers is not in the document]
Comment (roee88) [reply java/ffi/src/main/java/org/apache/arrow/ffi/SchemaExporter.java:66]: Feel free to commit directly with your suggested changes

**Pull request merged**